qiskit/providers/provider.py

```diff
@@ -75,3 +75,108 @@ def __eq__(self, other):
         """
         return type(self).__name__ == type(other).__name__
+
+
+class ProviderV2(Provider, ABC):
+    """Base class for a Backend Provider."""
+    version = 2
+
+    def get_backend(self, name=None, **kwargs):
+        """Return a single backend matching the specified filtering.
+
+        Args:
+            name (str): name of the backend.
+            **kwargs: dict used for filtering.
+
+        Returns:
+            Backend: a backend matching the filtering.
+
+        Raises:
+            QiskitBackendNotFoundError: if no backend could be found or
+                more than one backend matches the filtering criteria.
+        """
+        backends = self.backends(name, **kwargs)
+        if len(backends) > 1:
+            raise QiskitBackendNotFoundError('More than one backend matches the criteria')
+        if not backends:
+            raise QiskitBackendNotFoundError('No backend matches the criteria')
+
+        return backends[0]
+
+    @abstractmethod
+    @property
+    def backends(self):
+        """The backends provided by this provider.
+
+        This can be accessed by name via attributes for example::
+
+            my_provider = Provider()
+            my_provider.backend.backend_name
+
+        or called with the signature that matches
+        :meth:`~qiskit.providers.ProviderV1.backends`, for example::
+
+            my_provider = Provider()
+            my_provider.backend(name='backend_name')
+
+        Returns:
+            BackendsList: A :class:`~qiskit.providers.BackendsList` object
+        """
+        pass
+
+    def __eq__(self, other):
+        """Equality comparison.
+
+        By default, it is assumed that two `Providers` from the same class are
+        equal. Subclassed providers can override this behavior.
+        """
+        return type(self).__name__ == type(other).__name__
+
+
+class BackendList:
+    """A service class that allows for autocompletion of backends from a provider.
+
+    Each backend can be accessed as attribute by backend directly for example for two
+    backends ``BackendA`` (with a name ``'backend_a'``) and ``BackendB`` (with
+    a name ``'backend_b'``)::
+
+        from qiskit.providers import BackendV1
+
+        backends = BackendList([BackendA, BackendB])
+        backend_a = backends.backend_a.configuration()
+
+    would get the backend configuration object for backend_a. For backwards
+    compatibility a ``BackendList`` object is callable just as
+    :meth:`qiskit.providers.ProviderV1.backends`. For example::
+
+        backends = BackendList([BackendA, BackendB])
+        backend_list = backends(name='backend_a')
+
+    """
+
+    def __init__(self, backends):
+        """Initialize a new ``BackendList`` object.
+
+        Args:
+            backends (list): List of :class:`~qiskit.providers.Backend` instances.
+        """
+        self._backends = backends
+        for backend in backends:
+            setattr(self, backend.name(), backend)
```

Comment:
Not really a risk right now, but it's possible the backend name doesn't conform to Python identifier standard. ibmq has an `to_python_identifier` function to deal with that. If a user is using autocomplete, they should hopefully still recognize `My@Backend` is the same as `my_backend`.